File: Problem199_shortest-word-distance.java
// Time Complexity: O(n)
// Space Complexity: O(1)
class Solution {
    public int shortestDistance(String[] wordsDict, String word1, String word2) {
        int min = Integer.MAX_VALUE;
        int idxw1 = -1;
        int idxw2 = -1;
        for(int i = 0; i < wordsDict.length; i++) {
            if(word1.equals(wordsDict[i]))
                idxw1 = i;
            else if(word2.equals(wordsDict[i]))
                idxw2 = i;
            if(idxw1 != -1 && idxw2 != -1)
                min = Math.min(min, Math.abs(idxw1 - idxw2));
        }
        return min;
    }
}

File: Problem-199 Shortest Word Distance.py
# 243. Shortest Word Distance
# https://leetcode.com/problems/shortest-word-distance/

# Logic: We have 2 pointers, each for one of the two search words. We change 
# pointers whenever we find a match. We calculate the distance between the 
# pointers and update min.

# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def shortestDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:
        w1 = -1
        w2 = -1
        
        minDist = inf
        
        for idx, i in enumerate(wordsDict):
            if i == word1:
                w1 = idx
            elif i == word2:
                w2 = idx
            
            if w1 != -1 and w2 != -1:
                minDist = min(minDist, abs(w1-w2))
        
        return minDist